### PR TITLE
Feature/computed state

### PR DESCRIPTION
**Pull request by chandlerprall**
**NOTE: breaking change**

_Changes_
1. Breaks middleware, will need a major version bump and re-thinking how middleware works
2. Adds `store.addComputed(valueName, selectors, transformer)` (see [readme](https://github.com/chandlerprall/insula/blob/feature/computed-state/README.md#computed-values) for API)

_Edge cases considered but not directly handled_
1. Circular dependencies - I initially thought Insula would have to check that the `selectors` passed to `addComputed` do not create any circular dependency, as computed values can depend on computed values. Turns out this is not possible through the API as you have to create a computed value before you can reference it, avoiding circular dependencies entirely.
2. Determining/optimizing dependency execution order - expected to need code to determine the optimal order to execute nested computation methods, turns out Insula's structure does this already.
3. Executing all computed value subscriptions at once - expected to need code which would immediately execute affected computed-value-subscriptions instead of being queued in another batch and waiting for the next tick. Instead, Insula already executes these affected subscribers eagerly.

If anyone finds an execption / bug in the API for any of the above cases I will buy you lunch and/or beers.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -140,6 +141,38 @@ const unsubscribe = store.subscribeToState([['foo']], () => {});
...
+    user: {
+        title: 'Doctor',
+        firstName: 'Bobbiton',
+        lastName: 'Wobbiton',
+        suffix: 'Jr.'
+    }
+});
+
+const fullNameSelector = store.addComputed(
+    'fullName', // developer-friendly name for this value, used only for debugging
```
Comment (hartzis) [inline README.md]: mention this must be unique?

**Review by hartzis — APPROVED** [review]
Nice dude! amazing tests too!

**Comment by chandlerprall** [comment]
Waiting for #15 to be merged, there's probably a conflict and I'd much rather rebase this branch.

**Review by michaelmerrill — APPROVED** [review]
This looks great. Nice work!

**Pull request merged**